Jr/editor font size setting

jeremy-rifkin opened the pull request:
Implements #3641, at least the core functionality.

jeremy-rifkin commented on static/settings.ts:
```diff
@@ -269,9 +279,10 @@ export class Settings {
             name: keyof SiteSettings,
             populate: {label: string; desc: string}[],
-            defaultValue: string
+            defaultValue: any,
```
I'd like the type here to be `SiteSettings[name]` but I'm not sure if such a thing is possible in ts

RubenRBS replied:
Using generics it should be possible, but then I don't know if the function would still be usabe in this context. See https://www.typescriptlang.org/play?#code/C4TwDgpgBAyhzAJYDsDmBnKBeKBvAsAFBQlQCGAXFMgK4C2ARhAE4A0RpUDV6wzKqdsVIBjKgwD2EgDYQyyIQF8A3ESKhIsRMAhwEAzDgLCSyMnQhUA1hBASAZrHhI06IZwAmEe2RrTgAGpk0jSWTvquANo2do56LhgAukSKaoT2NMgiSBLI5B4ecLLZEswAPABy5tAQAB46yB6YMQ7hCegAfAAUZhZUVRasUF4+foHBoVTxBpEDEIkAlHhpqYREZAVFECXMXQBEZHtDAIwL65sQxcCl+4dDe7V7C6qEG4WX29e7eyJHUHyhZ7nd5XG4-P60aTSZ5AA

jeremy-rifkin replied:
Ah thank you! I think this can work

matsjla commented on static/colour.ts:
```diff
@@ -25,10 +25,12 @@
 import * as monaco from 'monaco-editor';
 import {Themes} from './themes';
 
+export type ColourScheme = 'rainbow' | 'rainbow2' | 'earth' | 'green-blue' | 'gray-shade' | 'rainbow-dark';
```
Good catch

matsjla commented on static/settings.ts:
```diff
@@ -265,13 +273,14 @@ export class Settings {
-        const addSelector = (
+        const addSelector = <Name extends keyof SiteSettings>(
             selector: string,
-            name: keyof SiteSettings,
+            name: Name,
             populate: {label: string; desc: string}[],
-            defaultValue: string
+            defaultValue: SiteSettings[Name],
+            component = Select
```
Great stuff, love the use of the Name generic here

jeremy-rifkin replied:
Thanks to Ruben for showing me how to make this work!

matsjla reviewed (approved):
LGTM - one thing I noticed is that the font size of existing editors doesn't change when the setting is changed, maybe this would be nice? I had to reset the layout for the change to apply

jeremy-rifkin commented:
> one thing I noticed is that the font size of existing editors doesn't change when the setting is changed, maybe this would be nice?

I was wondering about this too, I think I agree it'd be nice. I'll look into how to do it.

(One trick for now is just ctrl+click on all the font scale widgets to reset to the new default font scale)

jeremy-rifkin commented:
The default font scale setting change is now emitted as an event so all panes that listen for it will automatically update. I've implemented it for pane.ts and compiler/editor.js. More panes will get this automatically as they're all converted to ts. Some panes like ir-view aren't calling the base class's register callbacks method at the moment.

I also added a button here to apply your default to all current panes in the case of a share link
![image](https://user-images.githubusercontent.com/51220084/168743640-71b28c69-39c8-4733-847c-1aa43003144a.png)

jeremy-rifkin commented:
I'll merge in the morning if all looks good

jeremy-rifkin commented:
I'll merge in the morning if all looks good

partouf commented:
this is now live
